core/src/main/java/org/apache/lucene/analysis/miscellaneous/DuplicateByteSequenceSpotter.java

```diff
@@ -0,0 +1,318 @@
+        // non-sparse array.
+        TreeNode[] children;
+
+        RootTreeNode(byte key, TreeNode parentNode, int depth) {
+            super(key, parentNode, depth);
+            bytesAllocated += ROOT_TREE_NODE_OBJECT_SIZE;
+        }
+
+        public TreeNode add(byte b, int depth) {
+            if (children == null) {
+                children = new TreeNode[256];
+                bytesAllocated += (RamUsageEstimator.NUM_BYTES_OBJECT_REF * 256);
+            }
+            int bIndex = b + 128;
+            TreeNode node = children[bIndex];
+            if (node == null) {
+                if (depth <= 1) {
+                    // Depths 0 and 1 use RootTreeNode impl and create
+                    // RootTreeNodeImpl children
+                    node = new RootTreeNode(b, this, depth);
+                } else {
+                    // Deeper-level nodes are less visited but more numerous
+                    // so use a more space-friendly data structure
+                    node = new LightweightTreeNode(b, this, depth);
+                }
+                children[bIndex] = node;
+            }
+            return node;
+        }
+
+        @Override
+        public short add(int byteSequence) {
+            throw new UnsupportedOperationException("Root nodes do not support byte sequences encoded as integers");
+        }
+
+    }
+
+    // Node implementation for use by the depth 3 branches of the tree that
+    // sacrifices speed for space.
+    final class LightweightTreeNode extends TreeNode {
+
+        // An array dynamically resized but frequently only sized 1 as most 
+        // sequences leading to end leaves are one-off paths.
+        // It is scanned for matches sequentially and benchmarks showed
+        // that sorting contents on insertion didn't improve performance.
+        int[] children = null;
+
+        LightweightTreeNode(byte key, TreeNode parentNode, int depth) {
+            super(key, parentNode, depth);
+            bytesAllocated += LIGHTWEIGHT_TREE_NODE_OBJECT_SIZE;
+
+        }
+
+        @Override
+        public short add(int byteSequence) {
+            if (children == null) {
+                // Create array adding new child with the byte sequence combined with hitcount of 1.
+                // Most nodes at this level we expect to have only 1 child so we start with the  
+                // smallest possible child array.
+                children = new int[1];
+                bytesAllocated += RamUsageEstimator.NUM_BYTES_ARRAY_HEADER + Integer.BYTES;                
+                children[0] = byteSequence + 1;
+                return 1;
+            }
+            // Find existing child and if discovered increment count
+            for (int i = 0; i < children.length; i++) {
+                int child = children[i];
+                if (byteSequence == (child & 0xFFFFFF00)) {
+                    short hitCount = (short) (child & 0xFF);
```

Comment:
I'm not sure if the cast is worth it here. It is usually simpler to just work in integers even if we know if can't be more than 255.